File: bitchat/Views/Sonar/SonarOnboardingScreen.swift
//
// SonarOnboardingScreen.swift
// bitchat
//
// 3-step onboarding, ported from the Onboarding component in
// design/handoff/project/sonar/screens.jsx.
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

import SwiftUI

struct SonarOnboardingScreen: View {
    @EnvironmentObject private var store: SonarAppStore

    @State private var step = 0
    @State private var nick = ""
    @FocusState private var nickFocused: Bool
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private var can: Bool { nick.trimmingCharacters(in: .whitespaces).count >= 2 }
    private var trimmedNick: String { nick.trimmingCharacters(in: .whitespaces) }

    /// Nickname suggestions from the design handoff (data.js) — design copy,
    /// not demo data: tapping "Surprise me" picks one as the user's real nick.
    private static let suggestions = ["quietfox", "tram12", "lakeswim", "verdigris", "morningstatic", "papercrane", "northpine", "softsignal"]

    private func surprise() {
        nick = Self.suggestions.randomElement() ?? "quietfox"
    }

    var body: some View {
        VStack(spacing: 0) {
            // bc-obtop
            HStack {
                if step > 0 {
                    SNIconButton(action: { advance(to: step - 1) }) {
                        SNIcon(name: .back, size: 21, weight: 2.1)
                    }
                }
                Spacer()
            }
            .frame(height: 40)
            .padding(.leading, -10)

            // bc-obbody
            Group {
                switch step {
                case 0: stepIntro
                case 1: stepNickname
                default: stepDone
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
            .transition(reduceMotion ? .opacity : .offset(y: 10).combined(with: .opacity))
            .id(step)

            // bc-obfooter
            VStack(spacing: 16) {
                HStack(spacing: 6) {
                    ForEach(0..<3, id: \.self) { i in
                        Circle()
                            .fill(step == i ? SonarTheme.accent : SonarTheme.hairline)
                            .frame(width: 7, height: 7)
                    }
                }
                switch step {
                case 0:
                    SNPrimaryButton(label: "Get started") { advance(to: 1) }
                case 1:
                    SNPrimaryButton(label: "Continue", disabled: !can) { advance(to: 2) }
                default:
                    SNPrimaryButton(label: "Start chatting") {
                        store.completeOnboarding(nick: trimmedNick)
                    }
                }
            }
            .padding(.top, 18)
        }
        .padding(EdgeInsets(top: 10, leading: 28, bottom: 12, trailing: 28))
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(SonarTheme.bg.ignoresSafeArea())
    }

    private func advance(to newStep: Int) {
        if reduceMotion {
            step = newStep
        } else {
            withAnimation(.easeOut(duration: 0.35)) { step = newStep }
        }
    }

    // Step 1 — intro
    private var stepIntro: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            RoundedRectangle(cornerRadius: 23, style: .continuous)
                .fill(SonarTheme.accentFill)
                .frame(width: 74, height: 74)
                .overlay(
                    SNIcon(name: .rings, size: 40, weight: 1.5)
                        .foregroundColor(SonarTheme.onAccent)
                )
                .padding(.bottom, 28)
            Text("Sense who\u{2019}s nearby before you see them.")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("Sonar connects phones directly — no phone number, no account, no servers.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
                .padding(.bottom, 22)
            featureRow(icon: .mesh, title: "Works without internet", desc: "Bluetooth finds people around you, even offline.")
            featureRow(icon: .globe, title: "Out of range? Still reachable", desc: "Messages travel encrypted over the open internet instead.")
            featureRow(icon: .lock, title: "Private by design", desc: "Direct messages are end-to-end encrypted. Always.")
            Spacer()
        }
    }

    private func featureRow(icon: SNIconName, title: String, desc: String) -> some View {
        HStack(alignment: .top, spacing: 14) {
            RoundedRectangle(cornerRadius: 13, style: .continuous)
                .fill(SonarTheme.accentSoft)
                .frame(width: 40, height: 40)
                .overlay(
                    SNIcon(name: icon, size: 20)
                        .foregroundColor(SonarTheme.accentDeep)
                )
            VStack(alignment: .leading, spacing: 2) {
                Text(verbatim: title)
                    .font(SonarTheme.uiFont(size: 16, weight: .bold))
                    .foregroundColor(SonarTheme.text)
                Text(verbatim: desc)
                    .font(SonarTheme.uiFont(size: 13.5))
                    .lineSpacing(13.5 * 0.25)
                    .foregroundColor(SonarTheme.text2)
            }
        }
        .padding(.vertical, 11)
    }

    // Step 2 — nickname
    private var stepNickname: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            Text("Pick a nickname")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("It\u{2019}s just what people see — change it anytime.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
                .padding(.bottom, 22)
            HStack(spacing: 16) {
                SonarAvatar(name: trimmedNick.isEmpty ? "?" : trimmedNick, size: 72)
                TextField("", text: $nick, prompt: Text("nickname").foregroundColor(SonarTheme.text3).fontWeight(.medium))
                    .textFieldStyle(.plain)
                    .font(SonarTheme.uiFont(size: 21, weight: .bold))
                    .foregroundColor(SonarTheme.text)
                    .focused($nickFocused)
                    .onSubmit { if can { advance(to: 2) } }
                    .onChange(of: nick) { v in
                        if v.count > 20 { nick = String(v.prefix(20)) }
                    }
                    .padding(EdgeInsets(top: 15, leading: 16, bottom: 15, trailing: 16))
                    .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(SonarTheme.surface2))
                    .overlay(
                        RoundedRectangle(cornerRadius: 16, style: .continuous)
                            .strokeBorder(nickFocused ? SonarTheme.accent : Color.clear, lineWidth: 1.5)
                    )
            }
            .padding(.bottom, 18)
            Button(action: surprise) {
                HStack(spacing: 7) {
                    SNIcon(name: .dice, size: 16, weight: 2)
                    Text("Surprise me")
                        .font(SonarTheme.uiFont(size: 14, weight: .bold))
                }
                .foregroundColor(SonarTheme.accentDeep)
                .padding(EdgeInsets(top: 9, leading: 14, bottom: 9, trailing: 14))
                .background(Capsule().fill(SonarTheme.accentSoft))
            }
            .buttonStyle(SNScaleStyle(scale: 0.96))
            .padding(.top, 12)
            Text("No signup. Your identity is a private key created on this phone — nobody else ever sees it.")
                .font(SonarTheme.uiFont(size: 13))
                .lineSpacing(13 * 0.3)
                .foregroundColor(SonarTheme.text3)
                .padding(.top, 18)
            Spacer()
        }
    }

    // Step 3 — you're in
    private var stepDone: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            SonarAvatar(name: trimmedNick, size: 92)
                .padding(.bottom, 22)
            Text(verbatim: "You\u{2019}re in, \(trimmedNick).")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("No account was created anywhere — your identity lives on this phone.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
            SNFingerprintCard(label: "Your key fingerprint", value: store.myFingerprintDisplay)
                .padding(.top, 24)
            Text("Friends can verify this fingerprint in person to be sure it\u{2019}s really you.")
                .font(SonarTheme.uiFont(size: 13))
                .lineSpacing(13 * 0.3)
                .foregroundColor(SonarTheme.text3)
                .padding(.top, 18)
            Spacer()
        }
    }
}

/// bc-fpcard — fingerprint card used by onboarding step 3 and the profile screen.
struct SNFingerprintCard: View {
    let label: String
    let value: String

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(verbatim: label.uppercased())
                .font(SonarTheme.uiFont(size: 12, weight: .bold))
                .kerning(12 * 0.05)
                .foregroundColor(SonarTheme.text3)
            Text(verbatim: value)
                .font(SonarTheme.monoFont(size: 15))
                .kerning(15 * 0.08)
                .foregroundColor(SonarTheme.text)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(EdgeInsets(top: 13, leading: 16, bottom: 13, trailing: 16))
        .background(RoundedRectangle(cornerRadius: 14, style: .continuous).fill(SonarTheme.surface2))
    }
}

File: apps/sonar/composeApp/src/commonMain/kotlin/chat/bitchat/sonar/screens/SonarOnboardingScreen.kt
package chat.bitchat.sonar.screens

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.bitchat.sonar.SonarAppState
import chat.bitchat.sonar.ui.SNFingerprintCard
import chat.bitchat.sonar.ui.SNIcon
import chat.bitchat.sonar.ui.SNIconButton
import chat.bitchat.sonar.ui.SNIconName
import chat.bitchat.sonar.ui.SNPrimaryButton
import chat.bitchat.sonar.ui.SonarAvatar
import chat.bitchat.sonar.ui.sonar

private val SUGGESTIONS = listOf(
    "quietfox", "tram12", "lakeswim", "verdigris", "morningstatic", "papercrane", "northpine", "softsignal"
)

@Composable
fun SonarOnboardingScreen(state: SonarAppState) {
    val s = sonar
    var step by remember { mutableStateOf(0) }
    var nick by remember { mutableStateOf("") }
    val trimmed = nick.trim()
    val can = trimmed.length >= 2

    Column(
        Modifier.fillMaxSize().background(s.bg)
            .padding(start = 28.dp, end = 28.dp, top = 10.dp, bottom = 12.dp)
    ) {
        // Top bar (back)
        Box(Modifier.fillMaxWidth().height(40.dp)) {
            if (step > 0) {
                SNIconButton(SNIconName.Back, onClick = { step -= 1 })
            }
        }

        AnimatedContent(
            targetState = step,
            transitionSpec = { fadeIn() togetherWith fadeOut() },
            modifier = Modifier.weight(1f)
        ) { st ->
            when (st) {
                0 -> StepIntro()
                1 -> StepNickname(nick, trimmed) { nick = it.take(20) }
                else -> StepDone(trimmed, state.fingerprint())
            }
        }

        // Footer
        Column(Modifier.padding(top = 18.dp), horizontalAlignment = Alignment.CenterHorizontally) {
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                repeat(3) { i ->
                    Box(
                        Modifier.size(7.dp).clip(CircleShape)
                            .background(if (step == i) s.accent else s.hairline)
                    )
                }
            }
            Spacer(Modifier.height(16.dp))
            when (step) {
                0 -> SNPrimaryButton("Get started") { step = 1 }
                1 -> SNPrimaryButton("Continue", disabled = !can) { if (can) step = 2 }
                else -> SNPrimaryButton("Start chatting") { state.completeOnboarding(trimmed) }
            }
        }
    }
}

@Composable
private fun StepIntro() {
    val s = sonar
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        Box(
            Modifier.size(74.dp).clip(RoundedCornerShape(23.dp)).background(s.accentFill),
            contentAlignment = Alignment.Center
        ) { SNIcon(SNIconName.Rings, 40.dp, s.onAccent, weight = 1.5f) }
        Spacer(Modifier.height(28.dp))
        Text(
            "Sense who’s nearby before you see them.",
            color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black, lineHeight = 34.sp
        )
        Spacer(Modifier.height(10.dp))
        Text(
            "Sonar connects phones directly — no phone number, no account, no servers.",
            color = s.text2, fontSize = 16.sp, lineHeight = 21.sp
        )
        Spacer(Modifier.height(22.dp))
        FeatureRow(SNIconName.Mesh, "Works without internet", "Bluetooth finds people around you, even offline.")
        FeatureRow(SNIconName.Globe, "Out of range? Still reachable", "Messages travel encrypted over the open internet instead.")
        FeatureRow(SNIconName.Lock, "Private by design", "Direct messages are end-to-end encrypted. Always.")
    }
}

@Composable
private fun FeatureRow(icon: SNIconName, title: String, desc: String) {
    val s = sonar
    Row(Modifier.fillMaxWidth().padding(vertical = 11.dp)) {
        Box(
            Modifier.size(40.dp).clip(RoundedCornerShape(13.dp)).background(s.accentSoft),
            contentAlignment = Alignment.Center
        ) { SNIcon(icon, 20.dp, s.accentDeep) }
        Spacer(Modifier.width(14.dp))
        Column {
            Text(title, color = s.text, fontSize = 16.sp, fontWeight = FontWeight.Bold)
            Text(desc, color = s.text2, fontSize = 13.5.sp, lineHeight = 17.sp)
        }
    }
}

@Composable
private fun StepNickname(nick: String, trimmed: String, onChange: (String) -> Unit) {
    val s = sonar
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        Text("Pick a nickname", color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.height(10.dp))
        Text("It’s just what people see — change it anytime.", color = s.text2, fontSize = 16.sp, lineHeight = 21.sp)
        Spacer(Modifier.height(22.dp))
        Row(verticalAlignment = Alignment.CenterVertically) {
            SonarAvatar(trimmed.ifEmpty { "?" }, 72.dp)
            Spacer(Modifier.width(16.dp))
            Box(
                Modifier.weight(1f).clip(RoundedCornerShape(16.dp)).background(s.surface2)
                    .padding(horizontal = 16.dp, vertical = 15.dp)
            ) {
                if (nick.isEmpty()) Text("nickname", color = s.text3, fontSize = 21.sp, fontWeight = FontWeight.Medium)
                BasicTextField(
                    value = nick,
                    onValueChange = onChange,
                    singleLine = true,
                    textStyle = TextStyle(color = s.text, fontSize = 21.sp, fontWeight = FontWeight.Bold),
                    cursorBrush = SolidColor(s.accent),
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
        Spacer(Modifier.height(18.dp))
        Row(
            Modifier.clip(CircleShape).background(s.accentSoft).clickable {
                onChange(SUGGESTIONS[(nick.hashCode() and 0x7fffffff) % SUGGESTIONS.size])
            }.padding(horizontal = 14.dp, vertical = 9.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            SNIcon(SNIconName.Dice, 16.dp, s.accentDeep, weight = 2f)
            Spacer(Modifier.width(7.dp))
            Text("Surprise me", color = s.accentDeep, fontSize = 14.sp, fontWeight = FontWeight.Bold)
        }
        Spacer(Modifier.height(18.dp))
        Text(
            "No signup. Your identity is a private key created on this phone — nobody else ever sees it.",
            color = s.text3, fontSize = 13.sp, lineHeight = 17.sp
        )
    }
}

@Composable
private fun StepDone(nick: String, fingerprint: String) {
    val s = sonar
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        SonarAvatar(nick.ifEmpty { "?" }, 92.dp)
        Spacer(Modifier.height(22.dp))
        Text("You’re in, $nick.", color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.height(10.dp))
        Text("No account was created anywhere — your identity lives on this phone.", color = s.text2, fontSize = 16.sp, lineHeight = 21.sp)
        Spacer(Modifier.height(24.dp))
        SNFingerprintCard("Your key fingerprint", fingerprint.ifEmpty { "generating…" })
        Spacer(Modifier.height(18.dp))
        Text("Friends can verify this fingerprint in person to be sure it’s really you.", color = s.text3, fontSize = 13.sp, lineHeight = 17.sp)
    }
}
